doc/tdr.1
.ad b
.TH tdr 1 "25 June 1999" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.PP
.SH NAME
.PP
.I tdr
- test driver.
.PP
.SH SYNOPSIS
.PP
.I tdr
[-v [-i] ] [-o] 
.B cmd
[
.B args
]
.PP
.SH DESCRIPTION
.PP
.I tdr
is a program for testing programs.  
.I tdr
executes 
.B cmd
and
puts the various standard streams in the appropriate places.  
.B cmd
is
intended to be a here document with a .t extension.  If the extension
exists, then this part of the file name will be removed for the output
streams.  If there are expected output files present in the current
directory, then 
.I tdr
will do a 
.BR diff (1)
on the expected and actual
output and inform the user that differences were found.
.PP
.SH FILES
.PP
all these files should be found in the current directory the test program
is named 
.B fred
or 
.B fred.t
in this case:
.PP
.SH GENERATED FILES
.PP
   fred.ao - actual output created by 
.I tdr
   fred.aERR - actual error stream created by 
.I tdr
   fred.diff - differences (ie. unix diff) between
                  actual and expected output iff there
                  is a difference
   fred.diffERR - differences between actual and expected
                  error stream
   fred.aRET - text file with the return value
   tdr.log - log file for the test driver.
                  this file is appended to.
.PP
.SH EXISTING FILES
.PP
   fred.eo - expected output created by programmer
   fred.eERR - expected error stream created by programmer
   fred.eRET - expected return value
.PP
note that the diff files are only created if there is a difference.  the
program warns the user when a file is to be overwritten if -v option used
but doesn't prompt, unless the -i option is specified.
.PP
.SH OPTIONS
.PP
    -v warn before over writing files.
    -i prompt before over writing files.
    -o do not overwrite log file.
    -n print out handy statistical information (totals)
.PP
.SH RETURN VALUES
.PP
   0 - normal execution
   1 - user abort
   2 - usage line
   3 - file doesn't exist
   4 - difference between actual and expected output
.PP
.SH SEE ALSO
.PP
.BR diff (1)
.PP
.SH LEGAL
.PP
Copyright, 1996, Regents of University of California
.PP
portions are (c) 1997 Graham THE Ollis
.PP
.SH AUTHOR
.PP
Graham THE Ollis <ollisg@wwa.com>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


src/tdr.pl
#!/usr/bin/perl -w
##@(#)Copyright, 1996, Regents of University of California
##@(#)portions are (c) 1997 Graham THE Ollis
##==============================================================================
## test driver (tdr.pl)
## this program runs the given file and reports on
## how it worked out.  makes .ao (actual output) and
## compares it to .eo (expected output)
##
## History:
## Date       Author	Commented
## ----       ------	---------
## 20 May 96  G. Ollis	started & finished code
## 22 May 96  G. Ollis	added support for the ERR files
##			also made the code look nice
## 23 May 96  G. Ollis	added some command line options
##			-i for interactive
##			-v to include warnings
##			-s to display the signal sent to the test program
## 24 May 96  G. Ollis  trap the no command line arguments thing
##			check to see if the program is executable
## 29 May 96  G. Ollis  treat each argument as a command
##			rather than whole line is a cmdline
## 30 May 96  G. Ollis	one line output mode
## 3  Jun 96  G. Ollis	changed local() to my()
## 12 Jun 96  G. Ollis	changed $log back to local()
##			added accounting for the data
##			append to log file unless using the
##			fixed the code!  (it works this time)
## 2  Aug 96  G. Ollis	fixed a few little things, added documentation.
## 27 Feb 97  G. Ollis	return non zero if a difference is found.
##			added return value checking.
##			all the bases are covered in glue now, carefull you
##			don't fall down.
## 8  Oct 98  G. Ollis	added -n option to display passed/total stats after
##			testing
## 25 Jun 99  G. Ollis	once again, removed the documentation from here on a
##			whim into the netl dist.  this way i can use my new
##			rms format instead of the anoying POD format.
## 04 Jul 99  G. Ollis	added diffRET file.  it's kinda silly, but is better
##			than special casing it.
##==============================================================================
## see man page for more information.  (or POD below)

require 5;		# this funny thing doesn't even do anything but is a
			# great visual aid for the version impaired.

## GLOBALS
$VERSION	= '1.00';  if($VERSION) {}
$LOG_FILE_NAME	= 'tdr.log';		# where to put the test driver log
$DATE_CMD	= 'date | cut -c5-';	# date command

## options stuff
$interactive = 0;	# (-i) by default to non-interactive
$verbose = 0;		# (-v) by default don't print out warnings
$overWrite = 0;		# (-o) by default, DO NOT overwrite log file
$disp_stats = 0;	# (-n) by default do not print out handy stats
			# and only print out if a test case fails!

$return_value = 0;	# this is the value which will be returned
			# assuming it is a "normal" execution.

$0 =~ s!^.*/!!g;

unless(defined($ARGV[0])) {
	print STDERR "usage: $0 [-v  [-i] ] [-o] [-n] cmd [args]\n";
	exit 2;
}

while($ARGV[0] =~ /^-/) {
	$arg = shift(@ARGV);
	$interactive = 1 if($arg eq '-i');
	$verbose = 1 if($arg eq '-v');
	$overWrite = 1 if($arg eq '-o');
	$disp_stats = 1 if($arg eq '-n');
}

if((-e $LOG_FILE_NAME) and ($overWrite == 0)) {
	open(LOG,">>$LOG_FILE_NAME");
} else {
	open(LOG,">$LOG_FILE_NAME");
}

$total_work = 0;
$total = 0;
$total_na = 0;
for(@ARGV) {
	&testProgram($_);
}

if($disp_stats) {
	$percent = int($total_work / $total * 100);
	if($total == $total_work) {
		print "all $total tests passed\n";
	} elsif($total_na == 0) {
		print "$total_work/$total tests passed (${percent}%)\n";
		print $total - $total_work, " tests failed\n";
	} else {
		$total -= $total_na;
		print "$total_work/$total tests passed (${percent}%)\n";
		print $total - $total_work, " tests failed\n";
		print "no expected output for $total_na tests\n";
	}
}

exit $return_value;

##==============================================================================
## testProgram
## test the given program, make the right files
##
## INPUT : program name
## OUTPUT : debuging fles
##==============================================================================

#void
sub testProgram { # $progName
	$progName = $_[0];

	$total++;

	## if the name given doesn't have the .t extension, add it unless
	## there isn't a file without a .t extension the user wants to use.
	$progName .= ".t" unless(-x $progName);

	## if the file given or file.t doesn't exist, then tell the user
	## that he's going to have to do better.
	unless(-x $progName) {
		print "file doesn't exist, or permission denied.\n";
		exit 3;
	}

	## all the file names are given here relative to the command given to tdr
	my $root;	($root = $progName) =~ s!\.t$!!;
	my $fileName = "$root.ao";
	my $errName = "$root.aERR";
	my $oldFileName = "$root.eo";
	my $oldErrName = "$root.eERR";
	my $diffFileName = "$root.diff";
	my $diffErrName = "$root.diffERR";
	my $retName = "$root.aRET";
	my $oldRetName = "$root.eRET";
	my $diffRetName = "$root.diffRET";

	## warn the user that a file is to be overwriten
	&checkFile($fileName);
	&checkFile($errName);
	&checkFile($retName);

	## run the program, redirect as needed
	system "./$progName 1> $fileName 2> $errName";

	my $exitValue = ($? >> 8);
	my $signalValue = ($? & 255);
	# right shift 8 bits - the wonders of perl.  2 values in 16 bit int
	# $exitValue is the value returned by the program
	# $signalValue is the signal used to terminate it

	## save the return value for completeness
	open(RETURN, ">$retName") || die "could not open $retName for writing $!\n";
	print RETURN "exit($exitValue) signal($signalValue)\n";
	close RETURN;

	## if there is expected output warn of differences
	my $outDiff = &checkDiff($fileName, $oldFileName, $diffFileName);
	my $errDiff = &checkDiff($errName, $oldErrName, $diffErrName);
	my $retDiff = &checkDiff($retName, $oldRetName, $diffRetName);

	local $log;		# don't tell anyone i used a local here...

	if(($outDiff==2) or ($errDiff==2) or ($retDiff==2)) {	# diff FAILED!
		$log = "diff FAILED\n";    
	} elsif(($outDiff==3) or ($errDiff==3)) {		# no expected output
		$log = "N/A\n";
		$total_na++;

	} elsif(($outDiff==0) and ($errDiff==0) and 
				 (($retDiff==3) or ($retDiff==0)) ) {
		if($disp_stats) {
			$log = "OK\r";
		} else {
			$log = "OK\n";
		}
		$total_work++;
	} else {
		$log = '';
		$log .= 'STDOUT ' if $outDiff == 1;
		$log .= 'STDERR ' if $errDiff == 1;
		$log .= 'RETVAL ' if $retDiff == 1;
		$log .= "fail\n";
		$return_value = 4;
	}
	 
	write;
#	write LOG;

format STDOUT =
@<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< @<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
$progName,	$log
.

format LOG =
@<<<<<<<<<<<<<<<<<<<<<<< : @<<<<<<<<<<<<<< @*
&getTime,		    $progName,	    $log
.

}

##==============================================================================
## getTime
## get the time for logging purposes
##
## INPUT : none
## OUTPUT : the time/date fully qualified as a string
##==============================================================================

#string
sub getTime { # void
	my($fred);

	chop($fred = `$DATE_CMD`);

	$fred;
}

##==============================================================================
## checkDiff
## this function actually checks the difference between the two files given
## and put those differences in a .diff file
##
## INPUT : expected, actual, and diff file names
## OUTPUT : exit value of diff
##==============================================================================

#int
sub checkDiff { # $actualFileName, $expetedFileName, $diffFileName

	my($fileName, $oldFileName, $diffFileName) = @_;
	my $exitValue = 3;
	my $cmd = "diff $fileName $oldFileName";
	$cmd .= " 1> $diffFileName" if defined $diffFileName;

	if(-e $oldFileName) {  # if the expected output file exists, then....
		&checkFile($diffFileName);
		system $cmd;
		$exitValue = $? >> 8;
		unlink $fileName if $exitValue == 0;
		unlink $diffFileName if defined $diffFileName;
	}
	$exitValue;	# 0 = no difference
		# 1 = difference
		# 2 = diff failed
		# 3 = no expected output
}

##==============================================================================
## checkFile
## this function checks to see if the file can be over writen.
##
## INPUT : filename
## OUTPUT : none
##==============================================================================

#void 
sub checkFile { # $fileName
	my $fileName = $_[0];
	my $inp;

	return if defined $fileName and $fileName eq '/dev/null';

	return if($verbose == 0);
	if(-e $fileName) {
		print STDERR "warning, $fileName exists, ";
		if($interactive == 1) {
			print STDERR "overwrite? ";
			$inp = <STDIN>;
			exit 1 if($inp =~ /^n/);
		}
		print STDERR "overwriting.\n" if($interactive == 0);
	}
}

